Introduce --import-mode=importlib

nicoddemus opened the pull request:
This introduces --import-mode=importlib, which uses fine-grained facilities
from importlib to import test modules and conftest files, bypassing
the need to change sys.path and sys.modules as side-effect of that.

I've also opened #7245 to gather feedback on the new import mode.

Fix #5821

nicoddemus commented:
Unfortunately importing using `importlib` is bypassing our import hook (I updated a test specifically for that, which is failing).

Here's the code for `pyimport` which uses importlib:

https://github.com/pytest-dev/py/blob/f7b20d2cc71c96f2d0daba6ad2e9e83c49f60f1b/py/_path/local.py#L670-L685

Our own `AssertionRewritingHook` implements `importlib.abc.MetaPathFinder` and `importlib.abc.Loader`, so far as I understand things should "just work", but `AssertionRewritingHook.find_spec` is never called when using `importlib` to import test modules and conftest files.

@asottile, which is our in-house importlib expert, do you have any thoughts/hints here?

nicoddemus commented on testing/acceptance_test.py:
```diff
@@ -149,15 +149,16 @@ def my_dists():
 
-    def test_assertion_magic(self, testdir):
+    @pytest.mark.parametrize("import_mode", ["prepend", "append", "importlib"])
+    def test_assertion_rewrite(self, testdir, import_mode):
```
This is the test which fails with `"importlib"`.

asottile commented:
the difference in assertion rewriting comes down to the call to `importlib.util.spec_from_file_location` -- `py`lib is calling this with the default loader (`None`) which'll always result in `SourceFileLoader` (avoiding `sys.meta_path`) -- what we want is for that to pass the `loader=` argument with our assertion rewriter.  Here's an example from a debugging session where I used `pytest --import-mode=importlib t.py`.  Notice how I was able to retrieve the pytest rewriter from `sys.meta_path[0]` and construct a module spec with the right loader.

```
(Pdb) importlib.util.spec_from_file_location(modname, str(self))
ModuleSpec(name='t', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f94e6790ac0>, origin='/tmp/pytest/t.py')
(Pdb) importlib.util.spec_from_file_location(modname, str(self), loader=sys.meta_path[0])
ModuleSpec(name='t', loader=<_pytest.assertion.rewrite.AssertionRewritingHook object at 0x7f94e7023040>, origin='/tmp/pytest/t.py')
```

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -512,7 +512,7 @@ def _importconftest(self, conftestpath):
 
         try:
-            mod = conftestpath.pyimport()
+            mod = conftestpath.pyimport(ensuresyspath=importmode)
```
One question about this: previously this `pyimport` always used the default, even if the importmode was `append` (non-default). Now also `append` is plumbed into here, which changes the existing behavior. Does that make any difference?

nicoddemus replied:
It might change things indeed, but I'm inclined that this was actually an oversight.

What do you think @RonnyPfannschmidt?

nicoddemus commented:
@asottile:

>  what we want is for that to pass the loader= argument with our assertion rewriter.

Thanks, I was suspecting that.

But passing our assertion rewriter as loader won't interfere with other import hooks installed in `sys.meta_path`? I'm actually surprised `importlib` doesn't take in account `sys.meta_path` here at all.

If I'm right, should we explicitly loop over `sys.meta_path` instead?

asottile commented:
> 
> 
> @asottile:
> 
> > what we want is for that to pass the loader= argument with our assertion rewriter.
> 
> Thanks, I was suspecting that.
> 
> But passing our assertion rewriter as loader won't interfere with other import hooks installed in `sys.meta_path`? I'm actually surprised `importlib` doesn't take in account `sys.meta_path` here at all.
> 
> If I'm right, should we explicitly loop over `sys.meta_path` instead?

I'll have to think about this some more 🤔 but I suspect you're right that we'd need to implement the `sys.meta_path` lookup ourselves yeah

nicoddemus commented:
> I suspect you're right that we'd need to implement the sys.meta_path lookup ourselves yeah

How does this look like to you?

```python
import importlib.util
import sys


def test_foo():
    print()
    modname = "test_assert_foo"
    path = "test_assert_foo.py"
    for meta_hook in sys.meta_path:
        print('meta_hook=', meta_hook)
        x = meta_hook.find_module(modname, path)
        if x is not None:
            print('x=', x)
            spec = importlib.util.spec_from_file_location(modname, path, loader=meta_hook)
            print('spec=', spec)
            break

```

nicoddemus commented:
So I went ahead and my `sys.meta_path` handling above seems to work.

But I would need to change `py.path.local.pyimport`, but instead of doing that, decided to go ahead and copy it over to pytest as `_pytest.pathlib.import_module`. This is aligned with our desired to eventually remove `py` as a dependency, so it would have to be made eventually so I decided I might as well do it now. The tests for it can be improved in style and new tests should be added too, but I thought it better to first gather feedback here.

I suggest reviewing the individual commits, as there are a lot of changes I'm afraid.

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+        if modname is None:
+            modname = path.purebasename
+
+        for meta_importer in sys.meta_path:
```
@asottile here's the final code, let me know if it is correct. 👍

bluetech commented on doc/en/pythonpath.rst:
```diff
@@ -3,11 +3,65 @@
+Import modes
+------------
+
+pytest as a testing framework needs to import test moduels and ``conftest.py`` files for execution.
```
```suggestion
pytest as a testing framework needs to import test modules and ``conftest.py`` files for execution.
```

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    """ return path as an imported python module.
+
+    If modname is None, look for the containing package
+    and construct an according module name.
```
Some grammar error here, not sure what it means to say.

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    if it isn't already contained in sys.path.
+    if ensuresyspath is False no modification of syspath happens.
+
+    Special value of ensuresyspath=="importlib" is intended
```
I guess we should integrate this section into the above now, it's funny to read "purely for using in pytest" inside pytest :)

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
     return {sep.join(parts[: i + 1]) or sep for i in range(len(parts))}
+
+
+def import_module(path, modname=None, ensuresyspath=True):
```
Since `_pytest.pathlib` usually deals with `pathlib.Path`s, would be good to annotate `path: py.path.local` here to avoid confusion.

RonnyPfannschmidt replied:
lets annotate a 'pathLike or Path, and cast to Path if possible?

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    mild opt-in via this option. Note that it works only in
+    recent versions of python.
+    """
+    import py.error
```
How hard would it be to avoid `py.error` here? (Trying to reduce the `py` modules we directly depend on).

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+        modfile = mod.__file__
+        if modfile[-4:] in (".pyc", ".pyo"):
+            modfile = modfile[:-1]
+        elif modfile.endswith("$py.class"):
```
IIUC `$py.class` is a Jython thing, which we don't support (since it only supports Python 2), so can maybe remove this.

(If you prefer to keep this just a pure copy/paste that makes sense, can maybe make these changes later).

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+            try:
+                with open(str(path)) as f:
+                    exec(f.read(), mod.__dict__)
+            except:  # noqa: E722
```
```suggestion
            except BaseException:
```

asottile commented:
> 
> 
> > I suspect you're right that we'd need to implement the sys.meta_path lookup ourselves yeah
> 
> How does this look like to you?
> 
> ```python
> import importlib.util
> import sys
> 
> 
> def test_foo():
>     print()
>     modname = "test_assert_foo"
>     path = "test_assert_foo.py"
>     for meta_hook in sys.meta_path:
>         print('meta_hook=', meta_hook)
>         x = meta_hook.find_module(modname, path)
>         if x is not None:
>             print('x=', x)
>             spec = importlib.util.spec_from_file_location(modname, path, loader=meta_hook)
>             print('spec=', spec)
>             break
> ```

hmmm actually, if we're importing a test we always want to use the pytest loader right?  then we don't need to search through `sys.meta_path`

RonnyPfannschmidt commented on src/_pytest/pathlib.py:
```diff
@@ -365,3 +365,105 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    import py.error
+
+    if not path.check():
+        raise py.error.ENOENT(path)
```
`FileNotFoundError`

nicoddemus commented:
> hmmm actually, if we're importing a test we always want to use the pytest loader right? then we don't need to search through sys.meta_path

Not sure TBH... in `append` and `prepend` modes, other loaders are being considered by the high-level mechanism (AFAIU), so I don't know the consequences of ignoring them here. For example, I can think of a loader which adds itself to `sys.meta_path[0]` after pytest, but then uses pytest's loader to get assertion rewriting, and then does something else with the module. Plus seems "the right thing to do".

I might be misunderstanding how `importlib` works though.

> I get a couple of failures running pytest's own tests with --import-mode=importlib

Ahh that's a great idea, I should have done that already, thanks. I will take a look at those later.

Thanks about all the other comments regarding the `import_module`. I did notice many of the same problems mentioned while importing `pyimport`, but decided to postpone doing them until I got feedback that importing that function into pytest was the way to go.

So I understand I will continue on that route. 👍 

> im a bit onflicted on this one, the changes needed to port to pathlib are bi enough to warrnt a folloup instead of a "now"

I'm OK with doing them in this PR, actually. But of course I can do a follow up right after merging this if you folks prefer.

nicoddemus commented:
Did some cleanups/refactorings:

* Removed `import_module`'s `modname` paramter as pytest itself doesn't use it.
* Ported `ImportMismatchError` (as `ImportPathMismatchError` to avoid confusion)
* Use an enum `ImportMode` instead of magic `ensuresyspath=True/False/"importlib"`.
* Added more tests, including for `mode=ImportMode.importlib`.
* Ported `pypkpath` from `py` (as `resolve_package_path`)
* `import_module` now uses `Path` internally;

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    The import mechanism used is controlled by the `mode` parameter:
+
+    * `mode == ImportMode.prepend`: the directory containing the module (or package, taking
+      `__init__.py` files into account) will be put at the *start* of `sys.path`, if not
```
> if not already in  `sys.path`

According to the code, this is true for `append`, not `prepend`.

nicoddemus replied:
True, fixed. 👍

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+            sys.path.insert(0, str(pkgroot))
+
+    __import__(modname)
+
```
Is it intentional that the `sys.path` is not restored after the `__import__`?

nicoddemus replied:
Indeed it is; restoring `sys.path` later would cause surprising problems because of delayed imports: for example, a `conftest.py` file imported by this function might have local imports, which would fail at runtime if we restored `sys.path`.

bluetech replied:
Oo that's another good reason to move away from these modes.

A comment with your explanation would be useful!

nicoddemus replied:
Good point, done! 👍

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+
+    mod = sys.modules[modname]
+    if path.name == "__init__.py":
+        # we don't check anything as we might be in a namespace package...
```
This comment is not clear to me.

nicoddemus replied:
I ported it over from py, I'm removing it as I don't think it adds much

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+class ImportPathMismatchError(ImportError):
+    """Raised on import_module() if there is a mismatch of __file__'s.
+
+    This can happen when `import_module` is given different filenames with the same basename,
```
Indeed, let me try rephrasing it.

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    """
+
+
+def import_module(
```
Agreed, good idea. 👍

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    mode: Union[str, ImportMode] = ImportMode.prepend
+) -> ModuleType:
+    """
+    Imports and returns a module from the given path.
```
Indeed, updated docstring.

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+    pkgpath = resolve_package_path(path)
+    if pkgpath is not None:
+        pkgroot = pkgpath.parent
+        names = str(path.with_suffix("").relative_to(pkgroot)).split(os.path.sep)
```
Done thanks

nicoddemus commented on src/_pytest/pathlib.py:
```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+        if str(pkgroot) not in sys.path:
+            sys.path.append(str(pkgroot))
+    else:
+        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
```
Seems nice; I did implement it locally but I get:

```
src\_pytest\pathlib.py:495: error: Argument 1 to "assert_never" has incompatible type "ImportMode"; expected "NoReturn"  [arg-type]
Found 1 error in 1 file (checked 2 source files)
```

I implemented the solution from your original post verbatim, didn't really read the whole thread though.

bluetech replied:
Here's a diff that works for me. (Again, it's ok if you prefer not to apply this)

```diff
diff --git a/src/_pytest/compat.py b/src/_pytest/compat.py
index 84f9609a7..c19beb505 100644
--- a/src/_pytest/compat.py
+++ b/src/_pytest/compat.py
@@ -33,6 +33,7 @@ else:
 
 
 if TYPE_CHECKING:
+    from typing import NoReturn
     from typing import Type
     from typing_extensions import Final
 
@@ -401,3 +402,38 @@ else:
     from collections import OrderedDict
 
     order_preserving_dict = OrderedDict
+
+
+# Perform exhaustiveness checking.
+#
+# Consider this example:
+#
+#     MyUnion = Union[int, str]
+#
+#     def handle(x: MyUnion) -> int {
+#         if isinstance(x, int):
+#             return 1
+#         elif isinstance(x, str):
+#             return 2
+#         else:
+#             raise Exception('unreachable')
+#
+# Now suppose we add a new variant:
+#
+#     MyUnion = Union[int, str, bytes]
+#
+# After doing this, we must remember ourselves to go and update the handle
+# function to handle the new variant.
+#
+# With `assert_never` we can do better:
+#
+#     // throw new Error('unreachable');
+#     return assert_never(x)
+#
+# Now, if we forget to handle the new variant, the type-checker will emit a
+# compile-time error, instead of the runtime error we would have gotten
+# previously.
+#
+# This also work for Enums (if you use `is` to compare) and Literals.
+def assert_never(value: NoReturn) -> NoReturn:
+    assert False, "Unhandled value: {} ({})".format(value, type(value).__name__)
diff --git a/src/_pytest/pathlib.py b/src/_pytest/pathlib.py
index 907e9223d..da7f0221c 100644
--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -23,6 +23,7 @@ from typing import Union
 
 import py
 
+from _pytest.compat import assert_never
 from _pytest.warning_types import PytestWarning
 
 if sys.version_info[:2] >= (3, 6):
@@ -453,7 +454,7 @@ def import_path(
     if not path.exists():
         raise ImportError(path)
 
-    if mode == ImportMode.importlib:
+    if mode is ImportMode.importlib:
         import importlib.util
 
         module_name = path.stem
@@ -484,13 +485,14 @@ def import_path(
         pkg_root = path.parent
         module_name = path.stem
 
-    if mode == ImportMode.append:
+    if mode is ImportMode.append:
         if str(pkg_root) not in sys.path:
             sys.path.append(str(pkg_root))
-    else:
-        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
+    elif mode is ImportMode.prepend:
         if str(pkg_root) != sys.path[0]:
             sys.path.insert(0, str(pkg_root))
+    else:
+        assert_never(mode)
 
     __import__(module_name)
 
```

nicoddemus commented:
> Here's a diff that works for me.

No that's fine. Would you mind applying that and pushing to this branch? Thanks!

(I just think we need a `#pragma: no cover` in the `assert_never(mode)` line)

bluetech commented:
I would if I could figure out how to do so. Best I can do is a patch file you can apply with `git am`:

<details>

```diff
From 7092d42dbdf532fbcd98915bce1f6a48d7dd91d4 Mon Sep 17 00:00:00 2001
From: Ran Benita <ran@unusedvar.com>
Date: Mon, 8 Jun 2020 22:09:23 +0300
Subject: [PATCH] Check ImportMode exhaustively

---
 src/_pytest/compat.py  | 36 ++++++++++++++++++++++++++
 src/_pytest/pathlib.py | 10 ++++---
 testing/test_compat.py | 59 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 101 insertions(+), 4 deletions(-)

diff --git a/src/_pytest/compat.py b/src/_pytest/compat.py
index 84f9609a7..cd7dca719 100644
--- a/src/_pytest/compat.py
+++ b/src/_pytest/compat.py
@@ -33,6 +33,7 @@ else:
 
 
 if TYPE_CHECKING:
+    from typing import NoReturn
     from typing import Type
     from typing_extensions import Final
 
@@ -401,3 +402,38 @@ else:
     from collections import OrderedDict
 
     order_preserving_dict = OrderedDict
+
+
+# Perform exhaustiveness checking.
+#
+# Consider this example:
+#
+#     MyUnion = Union[int, str]
+#
+#     def handle(x: MyUnion) -> int {
+#         if isinstance(x, int):
+#             return 1
+#         elif isinstance(x, str):
+#             return 2
+#         else:
+#             raise Exception('unreachable')
+#
+# Now suppose we add a new variant:
+#
+#     MyUnion = Union[int, str, bytes]
+#
+# After doing this, we must remember ourselves to go and update the handle
+# function to handle the new variant.
+#
+# With `assert_never` we can do better:
+#
+#     // throw new Error('unreachable');
+#     return assert_never(x)
+#
+# Now, if we forget to handle the new variant, the type-checker will emit a
+# compile-time error, instead of the runtime error we would have gotten
+# previously.
+#
+# This also work for Enums (if you use `is` to compare) and Literals.
+def assert_never(value: "NoReturn") -> "NoReturn":
+    assert False, "Unhandled value: {} ({})".format(value, type(value).__name__)
diff --git a/src/_pytest/pathlib.py b/src/_pytest/pathlib.py
index 191018c04..afd2600d2 100644
--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -23,6 +23,7 @@ from typing import Union
 
 import py
 
+from _pytest.compat import assert_never
 from _pytest.outcomes import skip
 from _pytest.warning_types import PytestWarning
 
@@ -462,7 +463,7 @@ def import_path(
     if not path.exists():
         raise ImportError(path)
 
-    if mode == ImportMode.importlib:
+    if mode is ImportMode.importlib:
         import importlib.util
 
         module_name = path.stem
@@ -496,13 +497,14 @@ def import_path(
     # change sys.path permanently: restoring it at the end of this function would cause surprising
     # problems because of delayed imports: for example, a conftest.py file imported by this function
     # might have local imports, which would fail at runtime if we restored sys.path.
-    if mode == ImportMode.append:
+    if mode is ImportMode.append:
         if str(pkg_root) not in sys.path:
             sys.path.append(str(pkg_root))
-    else:
-        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
+    elif mode is ImportMode.prepend:
         if str(pkg_root) != sys.path[0]:
             sys.path.insert(0, str(pkg_root))
+    else:
+        assert_never(mode)
 
     __import__(module_name)
 
diff --git a/testing/test_compat.py b/testing/test_compat.py
index 45468b5f8..5debe87a3 100644
--- a/testing/test_compat.py
+++ b/testing/test_compat.py
@@ -1,16 +1,23 @@
+import enum
 import sys
 from functools import partial
 from functools import wraps
+from typing import Union
 
 import pytest
 from _pytest.compat import _PytestWrapper
+from _pytest.compat import assert_never
 from _pytest.compat import cached_property
 from _pytest.compat import get_real_func
 from _pytest.compat import is_generator
 from _pytest.compat import safe_getattr
 from _pytest.compat import safe_isclass
+from _pytest.compat import TYPE_CHECKING
 from _pytest.outcomes import OutcomeException
 
+if TYPE_CHECKING:
+    from typing_extensions import Literal
+
 
 def test_is_generator():
     def zap():
@@ -205,3 +212,55 @@ def test_cached_property() -> None:
     assert ncalls == 1
     assert c2.prop == 2
     assert c1.prop == 1
+
+
+def test_assert_never_union() -> None:
+    x = 10  # type: Union[int, str]
+
+    if isinstance(x, int):
+        pass
+    else:
+        with pytest.raises(AssertionError):
+            assert_never(x)  # type: ignore[arg-type]
+
+    if isinstance(x, int):
+        pass
+    elif isinstance(x, str):
+        pass
+    else:
+        assert_never(x)
+
+
+def test_assert_never_enum() -> None:
+    E = enum.Enum("E", "a b")
+    x = E.a  # type: E
+
+    if x is E.a:
+        pass
+    else:
+        with pytest.raises(AssertionError):
+            assert_never(x)  # type: ignore[arg-type]
+
+    if x is E.a:
+        pass
+    elif x is E.b:
+        pass
+    else:
+        assert_never(x)
+
+
+def test_assert_never_literal() -> None:
+    x = "a"  # type: Literal["a", "b"]
+
+    if x == "a":
+        pass
+    else:
+        with pytest.raises(AssertionError):
+            assert_never(x)  # type: ignore[arg-type]
+
+    if x == "a":
+        pass
+    elif x == "b":
+        pass
+    else:
+        assert_never(x)
-- 
2.26.2

```
</details>

nicoddemus commented:
I tried `git am`, but my latest changes don't let the patch apply cleanly anymore, sorry. 😬 

You can push to my branch like this:

```
$ git fetch git@github.com:nicoddemus/pytest.git import-mode-import-lib
$ git checkout -b nicoddemus/import-mode-import-lib FETCH_HEAD 
```

This will leave you in a branch `nicoddemus/import-mode-import-lib` with my branch changes.

After committing, you push with:

```
$ git push git@github.com:nicoddemus/pytest.git HEAD:import-mode-import-lib
```

nicoddemus commented:
Note: in 2a31c1a I'm used `importlib` also to import modules in `append` and `prepend` modes; previously it was using `__import__`.

nicoddemus commented:
> OK, pushed.

Thanks, appreciate it! 👍

nicoddemus commented:
Thanks everyone! 👍
